File: python/tvm/relay/op/vision/nms.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""Non-maximum suppression operations."""
from tvm.relay import expr
from . import _make


def get_valid_counts(data, score_threshold, id_index=0, score_index=1):
    """Get valid count of bounding boxes given a score threshold.
    Also moves valid boxes to the top of input data.

    Parameters
    ----------
    data : relay.Expr
        Input data. 3-D tensor with shape [batch_size, num_anchors, 6].

    score_threshold : optional, float
        Lower limit of score for valid bounding boxes.

    id_index : optional, int
        index of the class categories, -1 to disable.

    score_index: optional, int
        Index of the scores/confidence of boxes.

    Returns
    -------
    valid_count : relay.Expr
        1-D tensor for valid number of boxes.

    out_tensor : relay.Expr
        Rearranged data tensor.

    out_indices: relay.Expr
        Indices in input data
    """
    if not isinstance(score_threshold, expr.Expr):
        score_threshold = expr.const(score_threshold, "float32")
    return expr.TupleWrapper(
        _make.get_valid_counts(data, score_threshold, id_index, score_index), 3
    )


def non_max_suppression(
    data,
    valid_count,
    indices,
    max_output_size=-1,
    iou_threshold=0.5,
    force_suppress=False,
    top_k=-1,
    coord_start=2,
    score_index=1,
    id_index=0,
    return_indices=True,
    invalid_to_bottom=False,
):
    """Non-maximum suppression operator for object detection.

    Parameters
    ----------
    data : relay.Expr
        3-D tensor with shape [batch_size, num_anchors, 6]
        or [batch_size, num_anchors, 5].
        The last dimension should be in format of
        [class_id, score, box_left, box_top, box_right, box_bottom]
        or [score, box_left, box_top, box_right, box_bottom]. It could
        be the second output out_tensor of get_valid_counts.

    valid_count : relay.Expr
        1-D tensor for valid number of boxes. It could be the output
        valid_count of get_valid_counts.

    indices: relay.Expr
        2-D tensor with shape [batch_size, num_anchors], represents
        the index of box in original data. It could be the third
        output out_indices of get_valid_counts. The values in the
        second dimension are like the output of arange(num_anchors)
        if get_valid_counts is not used before non_max_suppression.

    max_output_size : int or relay.Expr, optional
        Max number of output valid boxes for each instance.
        Return all valid boxes if the value of max_output_size is less than 0.

    iou_threshold : float or relay.Expr, optional
        Non-maximum suppression threshold.

    force_suppress : bool, optional
        Suppress all detections regardless of class_id.

    top_k : int, optional
        Keep maximum top k detections before nms, -1 for no limit.

    coord_start : int, optional
        The starting index of the consecutive 4 coordinates.

    score_index : int, optional
        Index of the scores/confidence of boxes.

    id_index : int, optional
        index of the class categories, -1 to disable.

    return_indices : bool, optional
        Whether to return box indices in input data.

    invalid_to_bottom : bool, optional
        Whether to move all valid bounding boxes to the top.

    Returns
    -------
    out : relay.Expr or relay.Tuple
        return relay.Expr if return_indices is disabled, a 3-D tensor
        with shape [batch_size, num_anchors, 6] or [batch_size, num_anchors, 5].
        If return_indices is True, return relay.Tuple of two 2-D tensors, with
        shape [batch_size, num_anchors] and [batch_size, num_valid_anchors] respectively.
    """
    if not isinstance(max_output_size, expr.Expr):
        max_output_size = expr.const(max_output_size, "int32")
    if not isinstance(iou_threshold, expr.Expr):
        iou_threshold = expr.const(iou_threshold, "float32")
    out = _make.non_max_suppression(
        data,
        valid_count,
        indices,
        max_output_size,
        iou_threshold,
        force_suppress,
        top_k,
        coord_start,
        score_index,
        id_index,
        return_indices,
        invalid_to_bottom,
    )
    if return_indices:
        return expr.TupleWrapper(out, 2)
    return out


def all_class_non_max_suppression(
    boxes,
    scores,
    max_output_boxes_per_class=-1,
    iou_threshold=-1.0,
    score_threshold=-1.0,
    output_format="onnx",
):
    """Non-maximum suppression operator for object detection, corresponding to ONNX
    NonMaxSuppression and TensorFlow combined_non_max_suppression.
    NMS is performed for each class separately.

    Parameters
    ----------
    boxes : relay.Expr
        3-D tensor with shape (batch_size, num_boxes, 4)

    scores: relay.Expr
        3-D tensor with shape (batch_size, num_classes, num_boxes)

    max_output_boxes_per_class : int or relay.Expr, optional
        The maxinum number of output selected boxes per class

    iou_threshold : float or relay.Expr, optionaIl
        IoU test threshold

    score_threshold : float or relay.Expr, optional
        Score threshold to filter out low score boxes early

    output_format : string, optional
        "onnx" or "tensorflow". Specify by which frontends the outputs are
        intented to be consumed.

    Returns
    -------
    out : relay.Tuple
        If `output_format` is "onnx", the output is a relay.Tuple of two tensors, the first is
        `indices` of size `(batch_size * num_class* num_boxes , 3)` and the second is a scalar
        tensor `num_total_detection` of shape `(1,)` representing the total number of selected
        boxes. The three values in `indices` encode batch, class, and box indices.
        Rows of `indices` are ordered such that selected boxes from batch 0, class 0 come first,
        in descending of scores, followed by boxes from batch 0, class 1 etc. Out of
        `batch_size * num_class* num_boxes` rows of indices,  only the first `num_total_detection`
        rows are valid.

        If `output_format` is "tensorflow", the output is a relay.Tuple of three tensors, the first
        is `indices` of size `(batch_size, num_class * num_boxes , 2)`, the second is `scores` of
        size `(batch_size, num_class * num_boxes)`, and the third is `num_total_detection` of size
        `(batch_size,)` representing the total number of selected boxes per batch. The two values
        in `indices` encode class and box indices. Of num_class * num_boxes boxes in `indices` at
        batch b, only the first `num_total_detection[b]` entries are valid. The second axis of
        `indices` and `scores` are sorted within each class by box scores, but not across classes.
        So the box indices and scores for the class 0 come first in a sorted order, followed by
        the class 1 etc.
    """
    if not isinstance(max_output_boxes_per_class, expr.Expr):
        max_output_boxes_per_class = expr.const(max_output_boxes_per_class, "int32")
    if not isinstance(iou_threshold, expr.Expr):
        iou_threshold = expr.const(iou_threshold, "float32")
    if not isinstance(score_threshold, expr.Expr):
        score_threshold = expr.const(score_threshold, "float32")

    out = _make.all_class_non_max_suppression(
        boxes,
        scores,
        max_output_boxes_per_class,
        iou_threshold,
        score_threshold,
        output_format,
    )

    if output_format == "onnx":
        return expr.TupleWrapper(out, 2)

    return expr.TupleWrapper(out, 3)

File: src/relay/op/vision/nms.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/*!
 * \file nms.cc
 * \brief Non-maximum suppression operators
 */
#include <tvm/relay/attrs/vision.h>
#include <tvm/relay/op.h>
#include <tvm/tir/op.h>

namespace tvm {
namespace relay {

TVM_REGISTER_NODE_TYPE(GetValidCountsAttrs);

bool GetValidCountRel(const Array<Type>& types, int num_inputs, const Attrs& attrs,
                      const TypeReporter& reporter) {
  ICHECK_EQ(types.size(), 3);
  const auto* data = types[0].as<TensorTypeNode>();
  if (data == nullptr) return false;
  const auto& dshape = data->shape;
  ICHECK_EQ(dshape.size(), 3) << "Input data should be 3-D.";

  std::vector<IndexExpr> oshape({data->shape[0]});
  std::vector<IndexExpr> oshape_indices({data->shape[0], data->shape[1]});
  std::vector<Type> fields;
  fields.push_back(TensorType(oshape, DataType::Int(32)));
  fields.push_back(TensorType(data->shape, data->dtype));
  fields.push_back(TensorType(oshape_indices, DataType::Int(32)));

  // assign output type
  reporter->Assign(types[2], TupleType(Array<Type>(fields)));
  return true;
}

Expr MakeGetValidCounts(Expr data, Expr score_threshold, int id_index, int score_index) {
  auto attrs = make_object<GetValidCountsAttrs>();
  attrs->id_index = id_index;
  attrs->score_index = score_index;
  static const Op& op = Op::Get("vision.get_valid_counts");
  return Call(op, {data, score_threshold}, Attrs(attrs), {});
}

TVM_REGISTER_GLOBAL("relay.op.vision._make.get_valid_counts").set_body_typed(MakeGetValidCounts);

RELAY_REGISTER_OP("vision.get_valid_counts")
    .describe(R"doc(Get valid count of bounding boxes given
a score threshold. Also moves valid boxes to the top of
input data.
)doc" TVM_ADD_FILELINE)
    .set_num_inputs(2)
    .add_argument("data", "Tensor", "Input data.")
    .add_argument("score_threshold", "Tensor", "Minimum Score.")
    .set_support_level(5)
    .add_type_rel("GetValidCount", GetValidCountRel);

TVM_REGISTER_NODE_TYPE(NonMaximumSuppressionAttrs);

bool NMSRel(const Array<Type>& types, int num_inputs, const Attrs& attrs,
            const TypeReporter& reporter) {
  ICHECK_EQ(types.size(), 6);
  const auto* data = types[0].as<TensorTypeNode>();
  if (data == nullptr) return false;
  const auto* valid_count = types[1].as<TensorTypeNode>();
  if (valid_count == nullptr) return false;
  const NonMaximumSuppressionAttrs* param = attrs.as<NonMaximumSuppressionAttrs>();
  const auto& dshape = data->shape;
  const auto& vshape = valid_count->shape;
  ICHECK_EQ(dshape.size(), 3) << "Input data should be 3-D.";
  ICHECK_EQ(vshape.size(), 1) << "Input valid count should be 1-D.";

  // assign output type
  if (param->return_indices) {
    std::vector<Type> fields;
    // dynamic happens for return_indices in TensorFlow & ONNX
    std::vector<IndexExpr> oshape({dshape[0], dshape[1]});
    fields.push_back(TensorType(oshape, DataType::Int(32)));
    std::vector<IndexExpr> countshape({dshape[0], 1});
    fields.push_back(TensorType(countshape, DataType::Int(32)));
    reporter->Assign(types[5], TupleType(Array<Type>(fields)));
  } else {
    reporter->Assign(types[5], TensorType(dshape, data->dtype));
  }
  return true;
}

Expr MakeNMS(Expr data, Expr valid_count, Expr indices, Expr max_output_size, Expr iou_threshold,
             bool force_suppress, int top_k, int coord_start, int score_index, int id_index,
             bool return_indices, bool invalid_to_bottom) {
  auto attrs = make_object<NonMaximumSuppressionAttrs>();
  attrs->force_suppress = force_suppress;
  attrs->top_k = top_k;
  attrs->coord_start = coord_start;
  attrs->score_index = score_index;
  attrs->id_index = id_index;
  attrs->return_indices = return_indices;
  attrs->invalid_to_bottom = invalid_to_bottom;
  static const Op& op = Op::Get("vision.non_max_suppression");
  return Call(op, {data, valid_count, indices, max_output_size, iou_threshold}, Attrs(attrs), {});
}

TVM_REGISTER_GLOBAL("relay.op.vision._make.non_max_suppression").set_body_typed(MakeNMS);

RELAY_REGISTER_OP("vision.non_max_suppression")
    .describe(R"doc(Non-maximum suppression. The input boxes should
be in the format of [class_id, score, left, top, right, bottom]
or [score, left, top, right, bottom]. Set id_index to be -1 to
ignore class_id axis.
)doc" TVM_ADD_FILELINE)
    .set_num_inputs(5)
    .add_argument("data", "Tensor", "Input data.")
    .add_argument("valid_count", "Tensor", "Number of valid anchor boxes.")
    .add_argument("indices", "Tensor", "Corresponding indices in original input tensor.")
    .add_argument("max_output_size", "Tensor", "Max number of output valid boxes.")
    .add_argument("iou_threshold", "Tensor", "Threshold for box overlap.")
    .set_support_level(5)
    .add_type_rel("NMS", NMSRel);

TVM_REGISTER_NODE_TYPE(AllClassNonMaximumSuppressionAttrs);

bool AllClassNMSRel(const Array<Type>& types, int num_inputs, const Attrs& attrs,
                    const TypeReporter& reporter) {
  ICHECK_EQ(types.size(), 6);
  const auto* boxes = types[0].as<TensorTypeNode>();
  if (boxes == nullptr) return false;
  const auto* scores = types[1].as<TensorTypeNode>();
  if (scores == nullptr) return false;

  const auto& boxes_shape = boxes->shape;
  const auto& scores_shape = scores->shape;
  ICHECK_EQ(boxes_shape.size(), 3) << "Input boxes should be 3-D.";
  ICHECK_EQ(scores_shape.size(), 3) << "Input scores count should be 3-D.";

  IndexExpr batch = boxes_shape[0];
  IndexExpr num_classes = scores_shape[1];
  IndexExpr num_boxes = boxes_shape[1];

  const auto* param = attrs.as<AllClassNonMaximumSuppressionAttrs>();
  CHECK(param);

  std::vector<Type> fields;
  if (param->output_format == "onnx") {
    IndexExpr num_total_boxes = Any();
    if (!batch.as<AnyNode>() && !num_boxes.as<AnyNode>()) {
      num_total_boxes = batch * num_classes * num_boxes;
    }
    std::vector<IndexExpr> oshape{num_total_boxes, 3};
    std::vector<IndexExpr> counts_shape{1};
    fields.push_back(TensorType(oshape, DataType::Int(64)));
    fields.push_back(TensorType(counts_shape, DataType::Int(64)));
  } else {
    IndexExpr num_total_boxes_per_batch = Any();
    if (!num_boxes.as<AnyNode>()) {
      num_total_boxes_per_batch = num_classes * num_boxes;
    }
    std::vector<IndexExpr> indices_shape{batch, num_total_boxes_per_batch, 2};
    std::vector<IndexExpr> scores_shape{batch, num_total_boxes_per_batch};
    std::vector<IndexExpr> counts_shape{batch};
    fields.push_back(TensorType(indices_shape, DataType::Int(64)));
    fields.push_back(TensorType(scores_shape, DataType::Float(32)));
    fields.push_back(TensorType(counts_shape, DataType::Int(64)));
  }
  reporter->Assign(types[5], TupleType(Array<Type>(fields)));
  return true;
}

Expr MakeAllClassNMS(Expr boxes, Expr scores, Expr max_output_boxes_per_class, Expr iou_threshold,
                     Expr score_threshold, std::string output_format = "onnx") {
  auto attrs = make_object<AllClassNonMaximumSuppressionAttrs>();
  attrs->output_format = std::move(output_format);
  static const Op& op = Op::Get("vision.all_class_non_max_suppression");
  return Call(op, {boxes, scores, max_output_boxes_per_class, iou_threshold, score_threshold},
              Attrs(attrs), {});
}

TVM_REGISTER_GLOBAL("relay.op.vision._make.all_class_non_max_suppression")
    .set_body_typed(MakeAllClassNMS);

RELAY_REGISTER_OP("vision.all_class_non_max_suppression")
    .describe(R"doc(Non-maximum suppression operator for object detection, corresponding to ONNX
    NonMaxSuppression and TensorFlow combined_non_max_suppression.
    NMS is performed for each class separately
)doc" TVM_ADD_FILELINE)
    .set_num_inputs(5)
    .add_argument("boxes", "Tensor", "The input boxes in the format [batch, num_boxes, 4].")
    .add_argument("scores", "Tensor",
                  "Scores for each box and class in the format [batch, num_classes, num_boxes].")
    .add_argument("max_output_boxes_per_class", "Tensor",
                  "The maximum number of output boxes per class.")
    .add_argument("iou_threshold", "Tensor", "The IoU threshold for box the overlap test.")
    .add_argument("score_threshold", "Tensor",
                  "The score threshold to filter out low score boxes early.")
    .set_support_level(5)
    .add_type_rel("AllClassNMS", AllClassNMSRel);

TVM_REGISTER_NODE_TYPE(RegularNonMaximumSuppressionAttrs);

bool RegularNMSRel(const Array<Type>& types, int num_inputs, const Attrs& attrs,
                   const TypeReporter& reporter) {
  ICHECK_EQ(types.size(), 3);
  const auto* boxes = types[0].as<TensorTypeNode>();
  if (boxes == nullptr) return false;
  const auto* scores = types[1].as<TensorTypeNode>();
  if (scores == nullptr) return false;

  const auto& boxes_shape = boxes->shape;
  const auto& scores_shape = scores->shape;
  ICHECK_EQ(boxes_shape.size(), 3) << "Input boxes should be 3-D.";
  ICHECK_EQ(scores_shape.size(), 3) << "Input scores count should be 3-D.";

  IndexExpr num_batches = boxes_shape[0];

  const auto* param = attrs.as<RegularNonMaximumSuppressionAttrs>();
  CHECK(param);

  std::vector<Type> fields;
  std::vector<IndexExpr> nmsed_boxes_shape{num_batches, param->max_detections, 4};
  std::vector<IndexExpr> nmsed_classes_shape{num_batches, param->max_detections};
  std::vector<IndexExpr> nmsed_scores_shape{num_batches, param->max_detections};
  std::vector<IndexExpr> nmsed_detections_number_shape{num_batches};
  fields.push_back(TensorType(nmsed_boxes_shape, DataType::Float(32)));
  fields.push_back(TensorType(nmsed_classes_shape, DataType::Float(32)));
  fields.push_back(TensorType(nmsed_scores_shape, DataType::Float(32)));
  fields.push_back(TensorType(nmsed_detections_number_shape, DataType::Int(32)));

  reporter->Assign(types[2], TupleType(Array<Type>(fields)));
  return true;
}

Expr MakeRegularNMS(Expr boxes, Expr scores, int32_t max_detections_per_class,
                    int32_t max_detections, int32_t num_classes, double iou_threshold,
                    double score_threshold) {
  auto attrs = make_object<RegularNonMaximumSuppressionAttrs>();
  attrs->max_detections_per_class = max_detections_per_class;
  attrs->max_detections = max_detections;
  attrs->num_classes = num_classes;
  attrs->iou_threshold = iou_threshold;
  attrs->score_threshold = score_threshold;
  static const Op& op = Op::Get("vision.regular_non_max_suppression");
  return Call(op, {boxes, scores}, Attrs(attrs), {});
}

TVM_REGISTER_GLOBAL("relay.op.vision._make.regular_non_max_suppression")
    .set_body_typed(MakeRegularNMS);

RELAY_REGISTER_OP("vision.regular_non_max_suppression")
    .describe(R"doc(TBD)doc" TVM_ADD_FILELINE)
    .set_num_inputs(2)
    .add_argument("boxes", "Tensor",
                  "3-D tensor with shape (batch_size, num_boxes, 4). The four values in boxes "
                  "encode (ymin, xmin, ymax, xmax) coordinates of a box.")
    .add_argument("scores", "Tensor",
                  "3-D tensor with shape (batch_size, num_boxes, num_classes_with_background).")
    .set_support_level(5)
    .add_type_rel("RegularNMS", RegularNMSRel);

}  // namespace relay
}  // namespace tvm
